finesse/hardware/temperature/dp9800.py

```diff
@@ -0,0 +1,219 @@
+
+
+def check_data(data: bytes) -> None:
+    """Perform message integrity checks.
+
+    First check characters that we know should be constant.
+    Then calculate the BCC and compare it to that transmitted.
+
+    Args:
+        data: the message to check
+
+    Raises:
+        DP9800Error: Malformed message received from device
+    """
+    header_width = 2
+    value_width = 8
+    num_values = 9
+    if len(data) < (header_width + value_width * num_values):
+        raise DP9800Error("Insufficient data read")
+
+    if data[0] != 2:  # STX
+        raise DP9800Error("Start transmission character not detected")
+    if data.find(3) == -1:  # ETX
+        raise DP9800Error("End transmission character not detected")
+    if data[-1] != 0:  # NUL
+        raise DP9800Error("Null terminator not detected")
+
+    bcc = data[-2]
+    bcc_chars = data[1:-2]
+    byte_sum = 0
+    for byte in bcc_chars:
+        byte_sum ^= byte
+
+    if byte_sum != bcc:
+        raise DP9800Error("BCC check failed")
+
+
+def parse_data(data: bytes) -> tuple[list[Decimal], str]:
+    """Parse temperature data read from the DP9800.
+
+    The sequence of bytes is translated into a list of ASCII strings
+    representing each of the temperatures, and finally into floats.
+
+    Returns:
+        vals: A list of Decimals containing the temperature values recorded
+              by the DP9800 device.
+        sysflag: string representation of the system flag bitmask
+    """
+    try:
+        data_ascii = data.decode("ascii")
+    except Exception as e:
+        raise DP9800Error(e)
+
+    vals_begin = 3
+    vals_end = 74
+    etx_index = data.find(b"\x03")
+
+    vals = [Decimal(val) for val in data_ascii[vals_begin:vals_end].split()]
+
+    sysflag = bin(int(data_ascii[vals_end:etx_index], 16))
+
+    return (vals[1:], sysflag[2:])
+
+
+class DP9800Error(Exception):
+    """Indicates that an error occurred while communicating with the device."""
+
+
+class DP9800:
+    """An interface for DP9800 temperature readers.
+
+    The manual for this device is available at:
+    https://assets.omega.com/manuals/M5210.pdf
+    """
+
+    def __init__(self, serial: Serial) -> None:
+        """Create a new DP9800 from an existing serial device.
+
+        Args:
+            serial: Serial device
+        """
+        self.serial = serial
+
+        logging.info(f"Opened connection to DP9800 on port {self.serial.port}")
+        pub.sendMessage("temperature_monitor.open")
+        pub.subscribe(self.send_temperatures, "temperature_monitor.data.request")
+
+    def close(self) -> None:
+        """Close the connection to the device.
+
+        Raises:
+            SerialException: Error communicating with device
+        """
+        try:
+            self.serial.close()
+            pub.sendMessage("temperature_monitor.close")
+            logging.info("Closed connection to DP9800")
+        except SerialException as e:
+            raise DP9800Error(e)
```

Comment:
As you're only expecting to get exceptions on one line, I'd do this:

```suggestion
        try:
            self.serial.close()
        except SerialException as e:
            raise DP9800Error(e)
        else:
            pub.sendMessage("temperature_monitor.close")
            logging.info("Closed connection to DP9800")
```